src/net/socket_utils.zig

```diff
@@ -18,74 +19,344 @@ pub const PACKETS_PER_BATCH: usize = 64;
 // The identifier for the scoped logger used in this file.
 const LOG_SCOPE: []const u8 = "socket_utils";
 
-pub const SocketThread = struct {
-    allocator: Allocator,
-    handle: std.Thread,
+const XevThread = struct {
+    pub const Handle = std.Thread.ResetEvent;
 
-    pub fn spawnSender(
+    const List = std.DoublyLinkedList(struct {
         allocator: Allocator,
-        logger: Logger,
-        socket: UdpSocket,
-        outgoing_channel: *Channel(Packet),
-        exit: ExitCondition,
-    ) !*SocketThread {
-        return spawn(allocator, logger, socket, outgoing_channel, exit, runSender);
+        st: ?*SocketThread,
+        packet: Packet = undefined,
+        udp: xev.UDP,
+        udp_state: xev.UDP.State = undefined,
+        udp_completion: xev.Completion = undefined,
+        io: enum { idle, pending, cancelled } = .idle,
+    });
+
+    var ref_count = std.atomic.Value(usize).init(0);
+    var register = std.atomic.Value(?*List.Node).init(null);
+    var io_notified = std.atomic.Value(bool).init(false);
+    var io_thread: std.Thread = undefined;
+    var io_event: xev.Async = undefined;
+    var active = List{};
+
+    pub fn spawn(st: *SocketThread) !void {
+        // Make the socket non-blocking (required by xev).
+        var flags = try std.posix.fcntl(st.socket.internal, std.posix.F.GETFL, 0);
+        flags |= @as(c_int, @bitCast(std.posix.O{ .NONBLOCK = true }));
+        _ = try std.posix.fcntl(st.socket.internal, std.posix.F.SETFL, flags);
+
+        // Start xev thread if not running.
+        if (ref_count.fetchAdd(2, .acquire) == 0) {
```

Comment:
Is this meant to be 2?

Reply:
Seems intentional, would be good to comment why

Reply:
https://github.com/Syndica/sig/pull/465/commits/693219d352ad864800c5ad9da20211113ba6afab